engine/source_sym_engine.ml

```diff
@@ -61,122 +20,102 @@ let print_result stree =
+    let snd = relevant_wildcard_opt !with_wildcard
+    in
+    (fst, snd)
   in
-  let rec eval_clauses ?idx:(idx=0) constraints clauses : constraint_tree list =
-    match clauses with
-    | (pattern, exp)::tl ->
+  let rec decompose (rows: row list) : constraint_tree =
+    match rows with
+    | ((pattern::ptl), expr)::tl ->
       begin match pattern with
         | Or (p1, p2) ->
-          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
-        | Constructor (k, plist) ->
-          let pi = {idx=idx; op=piop_of_constructor k} in
-          let notpi = {pi with op=negate_piop pi.op} in 
-          let constraints' = (notpi :: constraints) in
-          begin match k with
-            | Int _ | String _ | Bool _ ->
-              assert (plist = []);
-              Leaf ([pi], exp)::(eval_clauses constraints' tl)
-            | Cons | Tuple -> assert (plist <> []);
-              let flat_children = plist |>
-                                  List.mapi (fun i pattern -> eval_clauses ~idx:i [] [(pattern, exp)])
-                                  |> BatList.n_cartesian_product
-                                  |> List.map (fun n -> Conjunction n)
-              in
-              flat_children @ eval_clauses constraints tl
-            | Variant _ when plist <> [] ->
-              let children = clauses
-                             |> List.filter_map (shares_kst ~wildcard_to_none:false k)
-                             |> eval_clauses []
-              in
-              let brothers = clauses
-                             |> List.filter (fun c -> shares_kst ~wildcard_to_none:true k c
-                                                      |> Option.is_none)
-                             |> eval_clauses constraints'
-              in
-              Node (pi, children) :: brothers
-            | Nil | Variant _ ->
-              assert (plist = []);
-              Leaf ([pi], exp) :: (eval_clauses constraints' tl)
-          end
-        | As (pattern, _) -> eval_clauses constraints ((pattern, exp)::tl)
-        | Wildcard ->
-          (* There could be a wildcard after another one, eg:
-           *   | K2 true -> ...
-           *   | K2 _ -> ...
-           *   | _ -> ...
-           * In that case we ignore it
-          *)
-          [Leaf (constraints, exp)]
+          decompose (((p1::ptl), expr) :: ((p2::ptl), expr) :: tl)
+        | Constructor (_, _) ->
+          let split, fallback = kst_rows_assoc rows
+          in
+          let split_evaluated =  split
+                                 |> List.map (fun (k, clause_lst) ->
+                                     k, decompose clause_lst)
+          in
+          let fallback_evaluated = match fallback with
+            | Some clause -> Some (decompose [clause])
+            | None -> None
+          in
+          Node (split_evaluated, fallback_evaluated)
+        | As (pattern, _) -> decompose (([pattern], expr)::tl)
+        | Wildcard -> ignore tl;
+          Leaf expr
```

Comment:
The `Wildcard` case is not correct anymore due to rows. The only `Leaf` cases should be those where the row is completely empty.

If the `kst_rows_assoc` is extended to deal with `As` and `Or` patterns correctly, then there is no need to do a matching on `pattern` here. You can call `kst_rows_assoc` in all cases, you get a group per constructor and a group of wildcards, and you evaluate each group recursively.